keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,428 @@
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+This feature would be able to implement by the same way to pv-protection/pvc-protection.
+- The deletion is blocked by using newly introduced `kubernetes.io/secret-protection` finalizer,
+- The finalizer will be added on creation of the secret by using admission controller,
+- The finalizer will be deleted by newly introduced `secret-protection-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resrouces.
+
+### Test Plan
+
+- For Alpha, unit tests and e2e tests verifying that a secret used by other resources is protected by this feature are added.
+- For Beta, scalability tests are added to exercise this feature.
+- For GA, the introduced e2e tests will be promoted to conformance.
+
+### Graduation Criteria
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- Allowing time for feedback
+
+#### Removing a Deprecated Flag
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+- Upgrade: Secret that doesn't have `kubernetes.io/secret-protection` finalizer will be added the finalizer on Update/Delete events, therefore no additional user operation will be needed.
+- Downgrade:
+  - Feature disabled case: If the secret-protection-controller exists and the feature is disabled, `kubernetes.io/secret-protection` finalizer will always be deleted, therefore no additional user operation will be needed,
+  - Downgraded to no secret-protection-controller case: If no secret-protection-controller exists but `kubernetes.io/secret-protection` finalizer is added to the secrets, no one remove the finalizer. Therefore, user needs to remove the `kubernetes.io/secret-protection` finalizer from all the secrets manually.
+
+### Version Skew Strategy
+
+- As for components, this feature involves only admission controller and secret-protection-controller, so version skew won't happen unless these components are available with different versions,
```

Comment:
Are these two proposed controllers in-tree or out-of-tree?  If in-tree, they can't watch VolumeSnapshot CRDs.

Reply:
Good question. As for the [PoC implementation](https://github.com/kubernetes/kubernetes/pull/101131), I used in-tree controller which covers only in-tree resources, Pod and PV. To add snapshot, we will need to add out-of-tree controller. These controllers should work independently if they use different finalizer.

I think that there will be a room of discussion on whether we should make the controller for in-tree resources to be an out-of-tree controller. Pv-protection/pvc-protection controllers is in-tree and secret-portection controller for snapshot must be out-of-tree, then where secret-portection controller for Pod and PV should exist?